Add AMA recap post to the blog (#106)
ethereum/solidity-blog: _posts/2020-11-04-solidity-ama-1-recap.md

One of the big advantages of the Yul-based optimizer is that each step can be seen in isolation: Each step receives Yul code as input and produces Yul code as output, without any tight dependency on other steps or analysis code. Furthermore, we try to keep each step as simple as possible so that bugs in those steps are very unlikely. As long as each simple step is bug-free, so is the whole Yul optimizer.

#### How does it work?

> There is single stage, the "constant optimizer" where the trade-off between deploy-time and run-time costs are taken into account. This stage tries to find a "better" representation of each number in the source, like ``0x10000000000`` can be encoded as ``PUSH6 0x10000000000`` (7 bytes and almost zero run-time costs), but it can also be encoded as `PUSH1 1 PUSH1 40 SHL` (5 bytes and a bit more expensive at run-time). Most of the time, the difference is not too relevant. The optimizer tries to simplify complicated expressions (which reduces both size and execution cost), but it also specializes or inlines functions. Especially function inlining is an operation that can cause much bigger code, but it is often done because it results in opportunities for more simplifications.

Author: For me this paragraph is quite hard to read and did not make sense as an introduction. Do we need it or is it ok to skip it? If we need it, we should try to reword it. Especially the "there is single stage" does not seem to make sense if you don't introduce / compare it to another stage. Which optimizer is this part about? (Or maybe I'm missing that, but it's definitely not a good introduction. Mabye we can move it as the second paragraph?)

> There is single stage, the "constant optimizer" where the trade-off between deploy-time and run-time costs are taken into account. This stage tries to find a "better" representation of each number in the source, like ``0x10000000000`` can be encoded as ``PUSH6 0x10000000000`` (7 bytes and almost zero run-time costs), but it can also be encoded as `PUSH1 1 PUSH1 40 SHL` (5 bytes and a bit more expensive at run-time). Most of the time, the difference is not too relevant. The optimizer tries to simplify complicated expressions (which reduces both size and execution cost), but it also specializes or inlines functions. Especially function inlining is an operation that can cause much bigger code, but it is often done because it results in opportunities for more simplifications.

Author: @chriseth what do you think about the above?

> There is single stage, the "constant optimizer" where the trade-off between deploy-time and run-time costs are taken into account. This stage tries to find a "better" representation of each number in the source, like ``0x10000000000`` can be encoded as ``PUSH6 0x10000000000`` (7 bytes and almost zero run-time costs), but it can also be encoded as `PUSH1 1 PUSH1 40 SHL` (5 bytes and a bit more expensive at run-time). Most of the time, the difference is not too relevant. The optimizer tries to simplify complicated expressions (which reduces both size and execution cost), but it also specializes or inlines functions. Especially function inlining is an operation that can cause much bigger code, but it is often done because it results in opportunities for more simplifications.

Reviewer: I think this is a useful information so we should keep it. Here are my two suggestions on how to improve it:
https://github.com/ethereum/solidity-blog/pull/106#discussion_r517479276
https://github.com/ethereum/solidity-blog/pull/106#discussion_r517484220

> There is single stage, the "constant optimizer" where the trade-off between deploy-time and run-time costs are taken into account. This stage tries to find a "better" representation of each number in the source, like ``0x10000000000`` can be encoded as ``PUSH6 0x10000000000`` (7 bytes and almost zero run-time costs), but it can also be encoded as `PUSH1 1 PUSH1 40 SHL` (5 bytes and a bit more expensive at run-time). Most of the time, the difference is not too relevant. The optimizer tries to simplify complicated expressions (which reduces both size and execution cost), but it also specializes or inlines functions. Especially function inlining is an operation that can cause much bigger code, but it is often done because it results in opportunities for more simplifications.

Author: Great, thank you!

In short, for the new code generator, one does not need to cache reads from a storage (or memory) if there are no writes to storage (or memory). Manual caching will only be beneficial in the following situation: if the loop contains a write, but if the contract author can reason that the write does not modify a variable that was read. An example of this situation would be the following:

    // Copying storage arrays arr1 into arr2, assuming arr2 is big enough.
    // Example where caching is helpful:
    // uint len = arr1.length
    // and replacing arr1.length with len will save gas

>      for(uint i = 0; i < arr1.length; ++i)

Reviewer: ```suggestion
    for(uint i = 0; i < arr1.length; ++i)
```

By the way, was there something wrong with the ` ``` ` blocks? I think they're easier to deal with than indentation :)

Reviewer 2: It may even be that ' ```solidity` provides better syntax highlighting

References:
- [High level description of steps in Yul optimizer](https://github.com/ethereum/solidity/blob/develop/libyul/optimiser/README.md)
- [List of steps in Yul optimizer](https://solidity.readthedocs.io/en/v0.7.4/yul.html?highlight=ir%20optimize#yul-optimizer)
- [Assembly based optimizer](https://solidity.readthedocs.io/en/v0.7.4/internals/optimiser.html)
<br><br>

#### What are the typical differences between optimized and non-optimized code?

Generally, the most visible difference would be constant expressions getting evaluated. When it comes to the ASM output, one can also notice reduction of equivalent/duplicate "code blocks" (compare the output of the flags `--asm` and `--asm --optimize`). However, when it comes to the Yul/intermediate-representation, there can be significant differences, for example, functions can get inlined, combined, rewritten to eliminate redundancies, etc. (compare the output between the flags `--ir` and `--optimize --ir-optimized`).

#### How is optimization affected by the number of runs (`--optimize-runs`) and is there a maximum number above which it stops mattering, or is `--optimize-runs=20000` less efficient than `--optimize-runs=500000`?

> The parameter specifies roughly how often each opcode of the deployed code will be executed across the life-time of the contract. A "runs" parameter of "1" will produce long but cheap code. The largest value is ``2**64-1``.

Reviewer 2: ```suggestion
The parameter specifies roughly how often each opcode of the deployed code will be executed across the life-time of the contract. A "runs" parameter of "1" will produce short but expensive code. The largest value is ``2**64-1``.
```

#### When multiple small variables are packed into a single storage slot, how can we be sure accessing them both is done with a single SSTORE/SLOAD? How "nearby" do the accesses have to be each-other, and is there any better way to ensure this than inspecting the assembly?

As long as this still uses the non-yul code generator, it is actually rather limiting. It is best to not have any branches in between the accesses, so assigning a memory struct to storage should work best.

#### Does it make sense to cache ``arr.length`` in a stack variable when looping over an array from front to back, or will it SLOAD/MLOAD that only once if it sees it won't change in the loop body?

> In the Yul optimizer, there is an optimization step called **LoopInvariantCodeMotion** designed to detect expressions that remain invariant in the loop and move them outside the loop. Take the following solidity example that finds the sum of a dynamic integer array.

Reviewer 2: We should stress that this section deals about the new code generator that is not active yet. For the old generator, there are some inefficiencies and it probably is cheaper to cache the length on the stack.

Reviewer 3: @franzihei The answer was slightly edited after posting to reddit. The text here is hackmd version https://www.reddit.com/r/ethdev/comments/jigz5o/ask_the_solidity_team_anything_1/gahzrhm/?utm_source=reddit&utm_medium=web2x&context=3 Can you please update it?

Reviewer 3: Here is the markdown source: https://gist.github.com/hrkrshnn/a1165fc31cbbf1fae9f271c73830fdda

Author: adjusted

### Solidity Optimizer

#### What is the Solidity Optimizer optimizing for (size or cost or something else?) and how does it achieve that?

The Solidity compiler uses two different optimizer modules: The "old" optimizer that operates at opcode level and the "new" optimizer that operates on Yul IR code. The opcode-based optimizer applies simplification rules from the [list](https://github.com/ethereum/solidity/blob/develop/libevmasm/RuleList.h) to opcodes next to each other. It also combines equal code sets, removes unused code and some other things. The Yul-based optimizer is much more powerful, because it can work across function calls: In Yul it is not possible to perform arbitrary jumps, so it is for example possible to compute the side-effects of each function: If a function does not modify storage, a call to it can be swapped with a function that does. If a function is side-effect free and its result is multiplied by zero, you can remove the function call completely.

> There is single stage, the "constant optimizer" where the trade-off between deploy-time and run-time costs are taken into account. This stage tries to find a "better" representation of each number in the source, like ``0x10000000000`` can be encoded as ``PUSH6 0x10000000000`` (7 bytes and almost zero run-time costs), but it can also be encoded as `PUSH1 1 PUSH1 40 SHL` (5 bytes and a bit more expensive at run-time). Most of the time, the difference is not too relevant. The optimizer tries to simplify complicated expressions (which reduces both size and execution cost), but it also specializes or inlines functions. Especially function inlining is an operation that can cause much bigger code, but it is often done because it results in opportunities for more simplifications.

Reviewer: Regarding https://github.com/ethereum/solidity-blog/pull/106#discussion_r517212279
> For me this paragraph is quite hard to read and did not make sense as an introduction. Do we need it or is it ok to skip it? If we need it, we should try to reword it. Especially the "there is single stage" does not seem to make sense if you don't introduce / compare it to another stage. Which optimizer is this part about? (Or maybe I'm missing that, but it's definitely not a good introduction. Mabye we can move it as the second paragraph?)

How about this:

```suggestion
In the opcode-based optimizer there is only one stage (the "constant optimizer") where the trade-off between deploy-time and run-time costs are taken into account. This stage tries to find a "better" representation of each number in the source, like ``0x10000000000`` can be encoded as ``PUSH6 0x10000000000`` (7 bytes and almost zero run-time costs), but it can also be encoded as `PUSH1 1 PUSH1 40 SHL` (5 bytes and a bit more expensive at run-time). Most of the time, the difference is not too relevant.

The optimizer tries to simplify complicated expressions (which reduces both size and execution cost), but it also specializes or inlines functions. Especially function inlining is an operation that can cause much bigger code, but it is often done because it results in opportunities for more simplifications.
```

I split the paragraph because I think the first part is only relevant for the opcode-based optimizer while the second one only for yul-optimizer (the old one does not do inlining).

One of the big advantages of the Yul-based optimizer is that each step can be seen in isolation: Each step receives Yul code as input and produces Yul code as output, without any tight dependency on other steps or analysis code. Furthermore, we try to keep each step as simple as possible so that bugs in those steps are very unlikely. As long as each simple step is bug-free, so is the whole Yul optimizer.

#### How does it work?

> There are several steps that the optimizer performs. A simple example would be evaluating expressions whose value is known at runtime, e.g., `x + 0` is evaluated to `x`, where `x` may be a parameter known only during runtime. A more complex example would be to identify expressions that remains invariant inside a loop and moving them outside the loop, thereby saving gas. Another interesting example would be avoiding accessing the same value from storage multiple times, i.e., multiple `sload` to the same slot can be reduced to a single `sload`, in certain situations.

Reviewer: I think this section is a good, high-level description of what optimizer does (both old and new one). I'd move it above "What is the Solidity Optimizer optimizing for".

Reviewer: Looks like you missed this one. Or do you think that it's better to leave it where it is?

Author: if we move this block up the "how does it work" section is empty, so I'd keep it here. :)

Reviewer: Yeah, but I meant the whole section. "how does it work" does sound like a more general question so it would make sense to have it higher to introduce the topic first.

As long as this still uses the non-yul code generator, it is actually rather limiting. It is best to not have any branches in between the accesses, so assigning a memory struct to storage should work best.

#### Does it make sense to cache ``arr.length`` in a stack variable when looping over an array from front to back, or will it SLOAD/MLOAD that only once if it sees it won't change in the loop body?

In the Yul optimizer, there is an optimization step called **LoopInvariantCodeMotion** designed to detect expressions that remain invariant in the loop and move them outside the loop. Take the following solidity example that finds the sum of a dynamic integer array.
 
```solidity
uint sum = 0;
for (uint i = 0; i < arr.length; ++i)
{
    sum += arr[i];
}
```
    
The optimization step can correctly identify that the `arr.length` remains invariant and will move it outside the loop. So there is no need for manually caching the length for this example.
    
To understand if you need to manually cache length, or any other storage/memory value inside a loop, we'll describe how the step works.
    

> 1. The step only deals with expressions that remains the same or invariant, so in the above example, `arr[i]` will not even be considered for moving.

Reviewer: ```suggestion
1. The step only deals with expressions that remain the same or invariant, so in the above example, `arr[i]` will not even be considered for moving.
```

2. If such expressions are movable, i.e., the expression does not have any side effects, they are moved outside right away. 
 Examples of such instructions would be arithmetic operations such as `add` or instructions that do not read/modify memory, storage or blockchain state, e.g., `address`. Non-examples would be `keccak256` (reads from memory), `sload` (reads from storage), `call` (can modify blockchain state and contract storage.)
3. If such expressions have side effects, but only the *read* kind, i.e., reading from storage or memory, e.g., `sload`, `mload`, `extcodesize`, etc., then they can be moved out of the loop if the loop does not write to the corresponding location.
In the above example, even though `arr.length` reads from storage, since no other expression in the loop can write to storage, we can move `arr.length` outside the loop.
    

>  In short, one does not need to cache reads from a storage (or memory) if there are no writes to storage (or memory.) Manual caching will only be beneficial in the following situation: if the loop contains a write, but if the contract author can reason that the write does not modify a variable that was read. An example of this situation would be the following:

Reviewer: ```suggestion
In short, one does not need to cache reads from a storage (or memory) if there are no writes to storage (or memory.) Manual caching will only be beneficial in the following situation: if the loop contains a write, but if the contract author can reason that the write does not modify a variable that was read. An example of this situation would be the following:
```

    
```solidity
// Copying arr1 into arr2, assuming arr2 is big enough.
// Example where caching is helpful:
// uint len = arr1.length
// and replacing arr1.length with len will save gas

> for(uint i = 0; i < arr1.length; ++i)

Reviewer: ```suggestion
for (uint i = 0; i < arr1.length; ++i)
```

{
    arr2[i] = arr1[i];
}
```
#### Does the compiler ever inline functions? Does it make sense for the user to be able to request the compiler to do this? Even if it's not worth it to reduce call overhead, this may allow new opportunities for constant folding.

The current code generator does not inline function, but the new one will, exactly for that purpose.

> 

Reviewer: I'd add this part too:

```markdown
We do not plan to make this user-controllable. It might also be that it makes sense to inline the function depending on how it is called, but in general, small functions are very likely to be inlined.
```

The Solidity compiler uses two different optimizer modules: The "old" optimizer that operates at opcode level and the "new" optimizer that operates on Yul IR code. The opcode-based optimizer applies simplification rules from the [list](https://github.com/ethereum/solidity/blob/develop/libevmasm/RuleList.h) to opcodes next to each other. It also combines equal code sets, removes unused code and some other things. The Yul-based optimizer is much more powerful, because it can work across function calls: In Yul it is not possible to perform arbitrary jumps, so it is for example possible to compute the side-effects of each function: If a function does not modify storage, a call to it can be swapped with a function that does. If a function is side-effect free and its result is multiplied by zero, you can remove the function call completely.

> In the opcode-based optimizer there is only one stage (the "constant optimizer") where the trade-off between deploy-time and run-time costs are taken into account. This stage tries to find a "better" representation of each number in the source, like ``0x10000000000`` can be encoded as ``PUSH6 0x10000000000`` (7 bytes and almost zero run-time costs), but it can also be encoded as `PUSH1 1 PUSH1 40 SHL` (5 bytes and a bit more expensive at run-time). Most of the time, the difference is not too relevant.
> 
> The optimizer tries to simplify complicated expressions (which reduces both size and execution cost), but it also specializes or inlines functions. Especially function inlining is an operation that can cause much bigger code, but it is often done because it results in opportunities for more simplifications.

Reviewer: Now that I look at it again, the second paragraph seems to make more sense on top, as a kind of introduction to this question.

I also added the missing `s` in `trade-offs`.

```suggestion
The optimizer tries to simplify complicated expressions (which reduces both size and execution cost), but it also specializes or inlines functions. Especially function inlining is an operation that can cause much bigger code, but it is often done because it results in opportunities for more simplifications.

In the opcode-based optimizer there is only one stage (the "constant optimizer") where the trade-offs between deploy-time and run-time costs are taken into account. This stage tries to find a "better" representation of each number in the source, like ``0x10000000000`` can be encoded as ``PUSH6 0x10000000000`` (7 bytes and almost zero run-time costs), but it can also be encoded as `PUSH1 1 PUSH1 40 SHL` (5 bytes and a bit more expensive at run-time). Most of the time, the difference is not too relevant.
```
